spec/app/ics-028-cross-chain-valset-update/README.md

```diff
@@ -0,0 +1,257 @@
+
+### Baby blockchain
+
+```golang
+// initialization function
+func init() {
+  UnbondingQueue = new local FIFO queue
+  validatorSetSeqNum = 0
+  dequeueSeqNum = 0
+  lastObserved = nil
+}
+```
+
+
+```golang
+// function used in End-Block method
+// returns all the elements enqueued since the last invocation of this method
+func observeChangesQ() { 
+  // get the content of the shared queue
+  content = Q.read()
+
+  // check whether the last observed element is in content
+  if (!content.has(lastObserved)) {
+    // update the lastObserved to the last element of content if content is not empty
+    if (!content.empty()) lastObserved = content.last()
+
+    // return the entire content
+    return content
+  } else {
+    // set "old" lastObserved
+    oldLastObserved = lastObserved
+
+    // update the lastObserved to the last element of content
+    lastObserved = content.last()
+
+    // return the entire content of content starting from oldLastObserved (excluding oldLastObserved)
+    return content.startFrom(oldLastObserved)
+  }
+}
+```
+
+- Expected precondition
+  - None
+- Expected postcondition
+  - Returns the elements enqueued since the last Q.read() operation was invoked.
+- Error condition
+  - None
+
+```golang
+// End-Block method executed at the end of each block
+func endBlock(block: Block) {
+  // get time
+  time = block.time
+
+  // finish unbonding for mature validator sets
+  while (!UnbondingQueue.isEmpty()) {
+    // peak the first queue entry
+    seqNum, startTime = UnbondingQueue.peak()
+
+    if (startTime + UNBONDING_PERIOD >= time) {
+      // remove from the unbonding queue
+      seqNum, startTime = UnbondingQueue.dequeue()
+
+      // dequeue all elements until seqNum
+      for (i = dequeueSeqNum + 1; i <= seqNum; i++) {
+        Q.dequeue()
+      }
+      dequeueSeqNum = seqNum
+
+    } else {
+      break
+    }
+  }
+
+  // get the new changes
+  changes = observeChangesQ()
+
+  // if there are no changes, return currnt validator set
+  if (changes.empty()) {
+    // validator set remains the same
+    ABCI.updateValidatorSet(block.validatorSet)
+    return
+  }
+
+  // get the old validator set
+  oldValSet = block.validatorSet
+
+  // get the new validator set; init to the old one
+  newValSet = oldValSet
+
+  // start unbonding for the old validator set represented by the validator set update;
+  // "start unbonding" simply means adding to the queue of validator set changes that started unbonding
+  UnbondingQueue.enqueue(validatorSetSeqNum, time)
+
+  // update the validator set
+  while (!changes.isEmpty()) {
+    valSetUpdate = changes.dequeue()
+
+    // update the new validator set
+    newValSet = applyValidatorUpdate(newValSet, valSetUpdate)
+
+    // increment validatorSetSeqNum
+    validatorSetSeqNum++
+
+    // remember which demands participate
+    if (content.isEmpty()) {
+      newSeqNum = validatorSetSeqNum
+    }
+  }
+
+  return newValSet
```

Comment:
Endblock just needs to return the `abci.ValidatorUpdate` not the entire validator set

Reply:
Fixed.